Migrate from UIImagePickerController to PHPickerViewController for image/video upload UI

rr-codes opened the pull request:
#### b82f73d063f37521fa94194d4cba86d0538c31df
<pre>
Migrate from UIImagePickerController to PHPickerViewController for image/video upload UI
<a href="https://bugs.webkit.org/show_bug.cgi?id=245906">https://bugs.webkit.org/show_bug.cgi?id=245906</a>
rdar://82449143

Reviewed by Aditya Keerthi.

Adopts the `PHPickerViewController` API for selecting media for file input fields
as `UIImagePickerController` is deprecated for the purpose of selecting media
from the Photo Library.

* Source/WebKit/Platform/spi/ios/PhotosUISPI.h:
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
(firstUTIThatConformsTo):
(-[WKFileUploadPanel presentWithParameters:resultListener:]):
(-[WKFileUploadPanel _dismissDisplayAnimated:]):
(-[WKFileUploadPanel contextMenuInteraction:configurationForMenuAtLocation:]):
(-[WKFileUploadPanel _showCamera]):
(-[WKFileUploadPanel _showPhotoPicker]):
(-[WKFileUploadPanel picker:didFinishPicking:]):
(-[WKFileUploadPanel _processPickerResults:successBlock:failureBlock:]):
(-[WKFileUploadPanel _processPickerResults:atIndex:processedResults:successBlock:failureBlock:]):
(-[WKFileUploadPanel _uploadItemFromResult:successBlock:failureBlock:]):
(-[WKFileUploadPanel _showPhotoPickerWithSourceType:]): Deleted.

Canonical link: <a href="https://commits.webkit.org/255435@main">https://commits.webkit.org/255435@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/97af1083c79a50b0cf70d3a32db017da6ae56d0a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92401 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1630 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22995 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102168 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1626 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30009 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84830 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98326 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98062 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1078 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78914 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28016 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/82996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82688 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/71111 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36424 "Built successfully") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16630 "Found 1 new test failure: webanimations/accelerated-web-animation-with-single-interval-and-easing-y-axis-above-1.html (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34192 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17801 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3769 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38063 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40417 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39966 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36952 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -353,6 +370,7 @@ @implementation WKFileUploadPanel {
     RetainPtr<WKFileUploadMediaTranscoder> _mediaTranscoder;
 #endif
     RetainPtr<UIImagePickerController> _imagePicker;
+    RetainPtr<PHPickerViewController> _photoPicker;
```
I really have no idea what I should name this

pxlcoder replied:
Maybe we rename `_imagePicker` since it's really just about the camera flow now.

pxlcoder commented on Source/WebKit/Platform/spi/ios/PhotosUISPI.h:
```diff
@@ -52,6 +56,20 @@
+
+@interface PHPickerResult : NSObject
+
+@end
```
`PHPickerViewController` and `PHPickerResult` are API, and we're not using any SPI from these classes, so they do not need to be in an SPI header.

pxlcoder commented on Source/WebKit/Platform/spi/ios/PhotosUISPI.h:
```diff
@@ -52,6 +56,20 @@
 
 @end
 
+@interface PHPickerConfiguration : NSObject
```
This should be a category rather than an interface declaration, and should be moved outside of `USE(APPLE_INTERNAL_SDK)`.

The purpose of declaring SPI is for open source builds where `USE(APPLE_INTERNAL_SDK)` is false.

rr-codes replied:
@pxlcoder do you have an example of a category that accomplishes something similar to what this is supposed to do?

pxlcoder replied:
There are many of them in `UIKitSPI.h` – `@interface UIDatePicker ()` is one example.

pxlcoder commented on Source/WebKit/Platform/spi/ios/PhotosUISPI.h:
```diff
@@ -52,6 +56,20 @@
 
+@interface PHPickerConfiguration : NSObject
+
+@property (nonatomic, setter=_setAllowsDownscaling:) BOOL _allowsDownscaling SPI_AVAILABLE(ios(16));
```
`SPI_AVAILABLE` shouldn't be added since this isn't our SPI.

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -33,6 +33,7 @@
 #import "APIOpenPanelParameters.h"
 #import "APIString.h"
 #import "PhotosUISPI.h"
+#import <PhotosUI/PHPicker.h>
```
Nit: Sort order. `<` comes after `"`.

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -59,9 +60,15 @@
 SOFT_LINK_CLASS(PhotosUIPrivate, PUActivityProgressController)
+SOFT_LINK_CLASS(PhotosUIPrivate, PHPickerConfiguration)
+SOFT_LINK_CLASS(PhotosUIPrivate, PHPickerViewController)
+SOFT_LINK_CLASS(PhotosUIPrivate, PHPickerResult)
```
None of these are in `PhotosUIPrivate`?

pxlcoder replied:
This comment still needs to be addressed. It's strange that we are using `PhotosUIPrivate` here, when all of these classes are public API, in `PhotosUI`.

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -815,6 +833,19 @@ - (void)_showPhotoPickerWithSourceType:(UIImagePickerControllerSourceType)source
+- (void)_showPhotoPicker
+{
+    auto config = adoptNS([allocPHPickerConfigurationInstance() init]);
+    config.get().selectionLimit = _allowMultipleFiles ? 0 : 1;
```
`[config setSelectionLimit:...]`

pxlcoder reviewed (commented):
I'm a little surprised there isn't more unused code that can be removed. Please see if there are any opportunities to do so.

Important things to consider:
- Is video encoded in the right format? I added custom logic to support this [here](https://bugs.webkit.org/show_bug.cgi?id=230639) and my understanding is the that new picker would do this for free.
- Is location metadata stripped from images?

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -993,6 +1049,75 @@ - (void)_processMediaInfoDictionaries:(NSArray *)infos atIndex:(NSUInteger)index
+- (void)_uploadItemFromResult:(PHPickerResult *)result successBlock:(void (^)(_WKFileUploadItem *))successBlock failureBlock:(void (^)(void))failureBlock
+{
+    if (NSString *uti = firstUTIThatConformsTo(result.itemProvider.registeredTypeIdentifiers, UTTypeMovie)) {
+        [result.itemProvider loadFileRepresentationForTypeIdentifier:uti completionHandler:^(NSURL *url, NSError *error) {
```
We will need to copy these files, as they will be destroyed by the time the data is returned to the web process.

pxlcoder commented on Source/WebKit/Platform/spi/ios/PhotosUISPI.h:
```diff
@@ -23,6 +23,7 @@
  * THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+#import <PhotosUI/PhotosUI.h>
```
Nit: Add a newline after this line.

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -43,6 +43,7 @@
 #import "WebIconUtilities.h"
 #import "WebOpenPanelResultListenerProxy.h"
 #import "WebPageProxy.h"
+#import <PhotosUI/PHPicker.h>
```
We could remove this since it will always be pulled in using `PhotosUISPI.h`.

pxlcoder commented on Source/WebKit/Shared/ios/WebIconUtilities.mm:
```diff
@@ -104,6 +106,7 @@ static CGRect squareCropRectForSize(CGSize size)
         (id)kCGImageSourceCreateThumbnailFromImageIfAbsent: @YES,
         (id)kCGImageSourceThumbnailMaxPixelSize: @(iconSideLength),
         (id)kCGImageSourceCreateThumbnailWithTransform: @YES,
+        (id)kCGImageSourceEnableRestrictedDecoding: @YES
```
cc @shallawa

shallawa replied:
Yes I think using `kCGImageSourceEnableRestrictedDecoding` is correct since it disables hardware decoding which may be affected by the sandbox. This is what we do when decoding images in WebCore.

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -816,6 +835,19 @@ - (void)_showPhotoPickerWithSourceType:(UIImagePickerControllerSourceType)source
 
+- (void)_showPhotoPicker
+{
+    auto config = adoptNS([allocPHPickerConfigurationInstance() init]);
```
Nit: `configuration`.

We generally prefer to avoid abbreviations / short forms in the code.

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -845,51 +877,28 @@ - (void)presentationControllerDidDismiss:(UIPresentationController *)presentatio
     [self _dismissDisplayAnimated:YES];
 
     dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), makeBlockPtr([retainedSelf = retainPtr(self), urlsFromUIKit = retainPtr(urlsFromUIKit)] () mutable {
+        ASSERT(!RunLoop::isMain());
```
This assertion isn't really useful, it's always under `dispatch_async(dispatch_get_global_queue`.

whsieh commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -902,6 +911,25 @@ - (void)documentPickerWasCancelled:(UIDocumentPickerViewController *)documentPic
+    [self _processPickerResults:results successBlock:^(NSArray<_WKFileUploadItem *> *items) {
+        ensureOnMainRunLoop([self, strongSelf = retainPtr(self), items = retainPtr(items)] {
+            [strongSelf->_view _removeTemporaryDirectoriesWhenDeallocated:WTFMove(strongSelf->_temporaryUploadedFileURLs)];
+            strongSelf->_temporaryUploadedFileURLs = { };
```
Nit - you can skip this if you've already moved `_temporaryUploadedFileURLs ` above. I would use `std::exchange(strongSelf->_temporaryUploadedFileURLs, { })` when calling `-_removeTemporaryDirectoriesWhenDeallocated:` as well, to make this more explicit.

rr-codes replied:
skip what exactly?

rr-codes replied:
do you suggest
```
[strongSelf->_view _removeTemporaryDirectoriesWhenDeallocated:WTFMove(std::exchange(strongSelf->_temporaryUploadedFileURLs, { }))];
```

instead of
```
[strongSelf->_view _removeTemporaryDirectoriesWhenDeallocated:WTFMove(strongSelf->_temporaryUploadedFileURLs)];
            strongSelf->_temporaryUploadedFileURLs = { };
```

?

whsieh replied:
My suggestion is:
```
[strongSelf->_view _removeTemporaryDirectoriesWhenDeallocated:std::exchange(strongSelf->_temporaryUploadedFileURLs, { })];
```

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -845,51 +877,28 @@ - (void)presentationControllerDidDismiss:(UIPresentationController *)presentatio
-        auto [maybeMovedURLs, temporaryURLs] = copyToNewTemporaryDirectory(urlsFromUIKit.get());
-        [retainedSelf->_view _removeTemporaryDirectoriesWhenDeallocated:WTFMove(temporaryURLs)];
+        [retainedSelf->_view _removeTemporaryDirectoriesWhenDeallocated:WTFMove(retainedSelf->_temporaryUploadedFileURLs)];
+        retainedSelf->_temporaryUploadedFileURLs = { };
```
This could use `std::exchange`.

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -902,6 +911,25 @@ - (void)documentPickerWasCancelled:(UIDocumentPickerViewController *)documentPic
+    [self _processPickerResults:results successBlock:^(NSArray<_WKFileUploadItem *> *items) {
+        ensureOnMainRunLoop([self, strongSelf = retainPtr(self), items = retainPtr(items)] {
+            [strongSelf->_view _removeTemporaryDirectoriesWhenDeallocated:WTFMove(strongSelf->_temporaryUploadedFileURLs)];
+            strongSelf->_temporaryUploadedFileURLs = { };
```
`std::exchange`

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -994,6 +1029,81 @@ - (void)_processMediaInfoDictionaries:(NSArray *)infos atIndex:(NSUInteger)index
+
+    auto uploadItemSuccessBlock = ^(_WKFileUploadItem *uploadItem) {
+        [processedResults addObject:uploadItem];
+
```
This newline can be removed.

whsieh commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -994,6 +1029,81 @@ - (void)_processMediaInfoDictionaries:(NSArray *)infos atIndex:(NSUInteger)index
+    auto uploadItemSuccessBlock = ^(_WKFileUploadItem *uploadItem) {
+        [processedResults addObject:uploadItem];
+
+        [self _processPickerResults:results atIndex:index manager:manager coordinator:coordinator processedResults:processedResults successBlock:successBlock failureBlock:failureBlock];
```
Hm...it's not super clear what keeps `manager` and `coordinator` alive in this block.

whsieh commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -974,6 +1022,89 @@ - (void)_processMediaInfoDictionaries:(NSArray *)infos successBlock:(void (^)(NS
+        }
+
+        auto [maybeMovedURL, temporaryURL] = [self _copyToNewTemporaryDirectory:url];
+        self->_temporaryUploadedFileURLs.append(WTFMove(temporaryURL));
```
Does anything guarantee that `self` is still alive when this block is called?

pxlcoder replied:
Doesn't the block implicitly retain `self`?

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -747,7 +780,10 @@ - (void)showFilePickerMenu
-    
+
+    _uploadFileManager = adoptNS([[NSFileManager alloc] init]);
+_uploadFileCoordinator = adoptNS([[NSFileCoordinator alloc] init]);
```
Indentation.

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -778,42 +811,60 @@ - (void)_adjustMediaCaptureType
+- (void)_showPhotoPicker
+{
+#if HAVE(PHOTOS_UI)
+    auto configuration = adoptNS([allocPHPickerConfigurationInstance() init]);
```
We should set `PHPickerConfigurationAssetRepresentationModeCompatible` on the configuration, for compatibility. The `Automatic` (default) value is subject to change.

If possible, we should also remove all the logic I added in https://bugs.webkit.org/show_bug.cgi?id=230639, if `PHPicker` does it for free.

rr-codes replied:
Isn't all that logic still needed for the camera case though?

pxlcoder replied:
That's something to test – the logic was originally added for multiple selection mode (see the Changelog).

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
```diff
@@ -778,42 +815,61 @@ - (void)_adjustMediaCaptureType
+#if HAVE(PHOTOS_UI)
+    auto configuration = adoptNS([allocPHPickerConfigurationInstance() init]);
+    [configuration setSelectionLimit:_allowMultipleFiles ? 0 : 1];
+    [configuration _setAllowsDownscaling:YES];
```
Nit: I would swap lines 861 and 862 for consistency (API/SPI usage).

pxlcoder commented on Source/WebKit/Shared/Cocoa/WebIconUtilities.mm:
```diff
@@ -116,6 +116,8 @@ PlatformImagePtr fallbackIconForFile(NSURL *file)
 #endif
 }
 
+const CFStringRef kCGImageSourceEnableRestrictedDecoding = CFSTR("kCGImageSourceEnableRestrictedDecoding");
```
Not new to your patch, and should be done separately, but we should make better use of `pal/spi/cg/ImageIOSPI.h`. `ImageDecoderCG` also defines a bunch of these.

rr-codes committed:
Migrate from UIImagePickerController to PHPickerViewController for image/video upload UI
https://bugs.webkit.org/show_bug.cgi?id=245906
rdar://82449143

Reviewed by Aditya Keerthi.

Adopts the `PHPickerViewController` API for selecting media for file input fields
as `UIImagePickerController` is deprecated for the purpose of selecting media
from the Photo Library.

* Source/WebKit/Platform/spi/ios/PhotosUISPI.h:
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
(firstUTIThatConformsTo):
(-[WKFileUploadPanel presentWithParameters:resultListener:]):
(-[WKFileUploadPanel _dismissDisplayAnimated:]):
(-[WKFileUploadPanel contextMenuInteraction:configurationForMenuAtLocation:]):
(-[WKFileUploadPanel _showCamera]):
(-[WKFileUploadPanel _showPhotoPicker]):
(-[WKFileUploadPanel picker:didFinishPicking:]):
(-[WKFileUploadPanel _processPickerResults:successBlock:failureBlock:]):
(-[WKFileUploadPanel _processPickerResults:atIndex:processedResults:successBlock:failureBlock:]):
(-[WKFileUploadPanel _uploadItemFromResult:successBlock:failureBlock:]):
(-[WKFileUploadPanel _showPhotoPickerWithSourceType:]): Deleted.

Canonical link: https://commits.webkit.org/255435@main
